cmake/functions.cmake

```diff
@@ -265,5 +303,43 @@ function(google_proto_library NAME)
     "${PROTOBUF_INCLUDE_DIR}")
   # TODO(hrapp): This should not explicityly list pthread and use
   # PROTOBUF_LIBRARIES, but that failed on first try.
-  target_link_libraries("${NAME}" "${PROTOBUF_LIBRARY}" pthread)
+  target_link_libraries("${NAME}" ${PROTOBUF_LIBRARY} pthread)
 endfunction()
+
+macro(google_initialize_cartographer_project)
+  SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules)
+  set(GOOG_CXX_FLAGS "-pthread -std=c++11 ${GOOG_CXX_FLAGS}")
+
+  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+    google_add_flag(GOOG_CXX_FLAGS "-Weverything")
+    google_add_flag(GOOG_CXX_FLAGS "-Werror=deprecated")
+    google_add_flag(GOOG_CXX_FLAGS "-Werror=non-pod-varargs")
+  else()
+    google_add_flag(GOOG_CXX_FLAGS "-Wall")
+    google_add_flag(GOOG_CXX_FLAGS "-Wpedantic")
+  endif()
+
+  # Turn some warnings into errors. These are defined for clang and gcc.
+  google_add_flag(GOOG_CXX_FLAGS "-Werror=format-security")
+  google_add_flag(GOOG_CXX_FLAGS "-Werror=return-type")
+  google_add_flag(GOOG_CXX_FLAGS "-Werror=uninitialized")
+
+  if(NOT CMAKE_BUILD_TYPE OR CMAKE_BUILD_TYPE STREQUAL "")
+    set(CMAKE_BUILD_TYPE Release)
+  endif()
+
+  if(CMAKE_BUILD_TYPE STREQUAL "Release")
+    google_add_flag(GOOG_CXX_FLAGS "-O3 -DNDEBUG")
+  elseif(CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
+    google_add_flag(GOOG_CXX_FLAGS "-O3 -g -DNDEBUG")
+  elseif(CMAKE_BUILD_TYPE STREQUAL "Debug")
+    message(FATAL_ERROR "Cartographer is too slow to be useful in debug mode.")
```

Comment:
Since this will be used in other projects, perhaps just say "Compiling in debug mode is not supported."